File: Samples/DeviceEnumerationAndPairing/cpp/DisplayHelpers.cpp
// Copyright (c) Microsoft. All rights reserved.

#include "pch.h"
#include "MainPage.xaml.h"
#include <ppltasks.h>
#include <DisplayHelpers.h>

using namespace SDKTemplate;
using namespace Platform;
using namespace Platform::Collections;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::UI::Core;
using namespace Windows::UI::Xaml::Data;
using namespace Windows::UI::Xaml::Interop;
using namespace Windows::UI::Xaml::Media::Imaging;
using namespace Windows::Devices::Enumeration;

// NOTE: The following using statements are only needed in order to demonstrate many of the
// different device selectors available from Windows Runtime APIs. You will only need to include
// the namespace for the Windows Runtime API your actual scenario needs.
using namespace Windows::Devices::HumanInterfaceDevice;
using namespace Windows::Networking::Proximity;
using namespace Windows::Devices::Bluetooth;
using namespace Windows::Devices::WiFiDirect;
using namespace Windows::Media::Casting;
using namespace Windows::Media::DialProtocol;
using namespace Windows::Devices::Sensors;
using namespace Windows::Devices::PointOfService;

namespace SDKTemplate
{
    // DeviceSelectorChoices

    Windows::Foundation::Collections::IVectorView<DeviceSelectorInfo^>^ DeviceSelectorChoices::CommonDeviceSelectors::get()
    {
        Vector<DeviceSelectorInfo^>^ selectors = ref new Vector<DeviceSelectorInfo^>();

        // Pre-canned device class selectors
        selectors->Append(ref new DeviceSelectorInfo("All Device Interfaces (default)", DeviceClass::All, nullptr));
        selectors->Append(ref new DeviceSelectorInfo("Audio Capture", DeviceClass::AudioCapture, nullptr));
        selectors->Append(ref new DeviceSelectorInfo("Audio Render", DeviceClass::AudioRender, nullptr));
        selectors->Append(ref new DeviceSelectorInfo("Image Scanner", DeviceClass::ImageScanner, nullptr));
        selectors->Append(ref new DeviceSelectorInfo("Location", DeviceClass::Location, nullptr));
        selectors->Append(ref new DeviceSelectorInfo("Portable Storage", DeviceClass::PortableStorageDevice, nullptr));
        selectors->Append(ref new DeviceSelectorInfo("Video Capture", DeviceClass::VideoCapture, nullptr));

        // A few examples of selectors built dynamically by windows runtime apis.
        selectors->Append(ref new DeviceSelectorInfo("Human Interface (HID)", DeviceClass::All, HidDevice::GetDeviceSelector(0, 0)));
        selectors->Append(ref new DeviceSelectorInfo("Activity Sensor", DeviceClass::All, ActivitySensor::GetDeviceSelector()));
        selectors->Append(ref new DeviceSelectorInfo("Pedometer", DeviceClass::All, Pedometer::GetDeviceSelector()));
        selectors->Append(ref new DeviceSelectorInfo("Proximity", DeviceClass::All, ProximityDevice::GetDeviceSelector()));
        selectors->Append(ref new DeviceSelectorInfo("Proximity Sensor", DeviceClass::All, ProximityDevice::GetDeviceSelector()));

        return selectors->GetView();
    }

    DeviceSelectorInfo^ DeviceSelectorChoices::Bluetooth::get()
    {
        // Currently Bluetooth APIs don't provide a selector to get ALL devices that are both paired and non-paired.  Typically you wouldn't need this for common scenarios, but it's convenient to demonstrate the
        // various sample scenarios. 
        return ref new DeviceSelectorInfo("Bluetooth", DeviceClass::All, "System.Devices.Aep.ProtocolId:=\"{e0cbf06c-cd8b-4647-bb8a-263b43f0f974}\"", DeviceInformationKind::AssociationEndpoint);
    }

    DeviceSelectorInfo^ DeviceSelectorChoices::BluetoothUnpairedOnly::get()
    {
        return ref new DeviceSelectorInfo("Bluetooth (unpaired)", DeviceClass::All, BluetoothDevice::GetDeviceSelectorFromPairingState(false));
    }

    DeviceSelectorInfo^ DeviceSelectorChoices::BluetoothPairedOnly::get()
    {
        return ref new DeviceSelectorInfo("Bluetooth (paired)", DeviceClass::All, BluetoothDevice::GetDeviceSelectorFromPairingState(true));
    }

    DeviceSelectorInfo^ DeviceSelectorChoices::BluetoothLE::get()
    {
        // Currently Bluetooth APIs don't provide a selector to get ALL devices that are both paired and non-paired.  Typically you wouldn't need this for common scenarios, but it's convenient to demonstrate the
        // various sample scenarios. 
        return ref new DeviceSelectorInfo("Bluetooth LE", DeviceClass::All, "System.Devices.Aep.ProtocolId:=\"{bb7bb05e-5972-42b5-94fc-76eaa7084d49}\"", DeviceInformationKind::AssociationEndpoint);
    }

    DeviceSelectorInfo^ DeviceSelectorChoices::BluetoothLEUnpairedOnly::get()
    {
        return ref new DeviceSelectorInfo("Bluetooth LE (unpaired)", DeviceClass::All, BluetoothLEDevice::GetDeviceSelectorFromPairingState(false));
    }

    DeviceSelectorInfo^ DeviceSelectorChoices::BluetoothLEPairedOnly::get()
    {
        return ref new DeviceSelectorInfo("Bluetooth LE (paired)", DeviceClass::All, BluetoothLEDevice::GetDeviceSelectorFromPairingState(true));
    }

    DeviceSelectorInfo^ DeviceSelectorChoices::WiFiDirect::get()
    {
        return ref new DeviceSelectorInfo("Wi-Fi Direct", DeviceClass::All, WiFiDirectDevice::GetDeviceSelector(WiFiDirectDeviceSelectorType::AssociationEndpoint));
    }

    DeviceSelectorInfo^ DeviceSelectorChoices::WiFiDirectPairedOnly::get()
    {
        return ref new DeviceSelectorInfo("Wi-Fi Direct (paired)", DeviceClass::All, WiFiDirectDevice::GetDeviceSelector());
    }

    DeviceSelectorInfo^ DeviceSelectorChoices::PointOfServicePrinter::get()
    {
        return ref new DeviceSelectorInfo("Point of Service Printer", DeviceClass::All, PosPrinter::GetDeviceSelector());
    }

    DeviceSelectorInfo^ DeviceSelectorChoices::VideoCasting::get()
    {
        return ref new DeviceSelectorInfo("Video Casting", DeviceClass::All, CastingDevice::GetDeviceSelector(CastingPlaybackTypes::Video));
    }

    DeviceSelectorInfo^ DeviceSelectorChoices::DialAllApps::get()
    {
        return ref new DeviceSelectorInfo("DIAL (All apps)", DeviceClass::All, DialDevice::GetDeviceSelector(""));
    }

    // WSD and UPnP are unique in that there are currently no general WSD or UPnP APIs to USE the devices once you've discovered them. 
    // You can pair the devices using DeviceInformation.Pairing.PairAsync etc, and you can USE them with the sockets APIs. However, since
    // there's no specific API right now, there's no *.GetDeviceSelector available.  That's why we just simply build the device selector
    // ourselves and specify the correct DeviceInformationKind (AssociationEndpoint). 
    DeviceSelectorInfo^ DeviceSelectorChoices::Wsd::get()
    {
        return ref new DeviceSelectorInfo("Web Services on Devices (WSD)", DeviceClass::All, "System.Devices.Aep.ProtocolId:=\"{782232aa-a2f9-4993-971b-aedc551346b0}\"", DeviceInformationKind::AssociationEndpoint);
    }

    DeviceSelectorInfo^ DeviceSelectorChoices::Upnp::get()
    {
        return ref new DeviceSelectorInfo("UPnP", DeviceClass::All, "System.Devices.Aep.ProtocolId:=\"{0e261de4-12f0-46e6-91ba-428607ccef64}\"", DeviceInformationKind::AssociationEndpoint);
    }


    IVectorView<DeviceSelectorInfo^>^ DeviceSelectorChoices::DevicePickerSelectors::get()
    {
        Vector<DeviceSelectorInfo^>^ selectors = ref new Vector<DeviceSelectorInfo^>(begin(CommonDeviceSelectors), end(CommonDeviceSelectors));
        selectors->Append(BluetoothPairedOnly);
        selectors->Append(BluetoothUnpairedOnly);
        selectors->Append(BluetoothLEPairedOnly);
        selectors->Append(BluetoothLEUnpairedOnly);
        selectors->Append(WiFiDirect);
        selectors->Append(PointOfServicePrinter);
        selectors->Append(VideoCasting);
        selectors->Append(DialAllApps);

        return selectors->GetView();
    }

    IVectorView<DeviceSelectorInfo^>^ DeviceSelectorChoices::FindAllAsyncSelectors::get()
    {
        Vector<DeviceSelectorInfo^>^ selectors = ref new Vector<DeviceSelectorInfo^>(begin(CommonDeviceSelectors), end(CommonDeviceSelectors));
        selectors->Append(BluetoothPairedOnly);
        selectors->Append(BluetoothLEPairedOnly);
        selectors->Append(WiFiDirectPairedOnly);

        return selectors->GetView();
    }

    IVectorView<DeviceSelectorInfo^>^ DeviceSelectorChoices::DeviceWatcherSelectors::get()
    {
        Vector<DeviceSelectorInfo^>^ selectors = ref new Vector<DeviceSelectorInfo^>(begin(CommonDeviceSelectors), end(CommonDeviceSelectors));
        selectors->Append(Bluetooth);
        selectors->Append(BluetoothLE);
        selectors->Append(WiFiDirect);
        selectors->Append(PointOfServicePrinter);
        selectors->Append(VideoCasting);
        selectors->Append(DialAllApps);
        selectors->Append(Wsd);
        selectors->Append(Upnp);

        return selectors->GetView();
    }

    IVectorView<DeviceSelectorInfo^>^ DeviceSelectorChoices::BackgroundDeviceWatcherSelectors::get()
    {
        Vector<DeviceSelectorInfo^>^ selectors = ref new Vector<DeviceSelectorInfo^>(begin(CommonDeviceSelectors), end(CommonDeviceSelectors));
        selectors->Append(BluetoothPairedOnly);
        selectors->Append(BluetoothLEPairedOnly);
        selectors->Append(WiFiDirectPairedOnly);
        selectors->Append(PointOfServicePrinter);
        selectors->Append(VideoCasting);
        selectors->Append(DialAllApps);
        selectors->Append(Wsd);
        selectors->Append(Upnp);

        return selectors->GetView();
    }

    IVectorView<DeviceSelectorInfo^>^ DeviceSelectorChoices::PairingSelectors::get()
    {
        Vector<DeviceSelectorInfo^>^ selectors = ref new Vector<DeviceSelectorInfo^>();
        selectors->Append(Bluetooth);
        selectors->Append(BluetoothLE);
        selectors->Append(WiFiDirect);
        selectors->Append(PointOfServicePrinter);
        selectors->Append(VideoCasting);
        selectors->Append(Wsd);
        selectors->Append(Upnp);

        return selectors->GetView();
    }

    // DeviceInformationKindChoices

    IVectorView<DeviceInformationKindChoice^>^ DeviceInformationKindChoices::Choices::get()
    {
        Vector<DeviceInformationKindChoice^>^ choices = ref new Vector<DeviceInformationKindChoice^>();

        Vector<DeviceInformationKind>^ kinds;
        
        kinds = ref new Vector<DeviceInformationKind>();
        kinds->Append(DeviceInformationKind::DeviceContainer);
        choices->Append(ref new DeviceInformationKindChoice("DeviceContainer", kinds->GetView()));

        kinds = ref new Vector<DeviceInformationKind>();
        kinds->Append(DeviceInformationKind::Device);
        choices->Append(ref new DeviceInformationKindChoice("Device", kinds->GetView()));

        kinds = ref new Vector<DeviceInformationKind>();
        kinds->Append(DeviceInformationKind::DeviceInterface);
        choices->Append(ref new DeviceInformationKindChoice("DeviceInterface", kinds->GetView()));

        kinds = ref new Vector<DeviceInformationKind>();
        kinds->Append(DeviceInformationKind::DeviceInterfaceClass);
        choices->Append(ref new DeviceInformationKindChoice("DeviceInterfaceClass", kinds->GetView()));

        kinds = ref new Vector<DeviceInformationKind>();
        kinds->Append(DeviceInformationKind::AssociationEndpointContainer);
        choices->Append(ref new DeviceInformationKindChoice("AssociationEndpointContainer", kinds->GetView()));

        kinds = ref new Vector<DeviceInformationKind>();
        kinds->Append(DeviceInformationKind::AssociationEndpoint);
        choices->Append(ref new DeviceInformationKindChoice("AssociationEndpoint", kinds->GetView()));

        kinds = ref new Vector<DeviceInformationKind>();
        kinds->Append(DeviceInformationKind::AssociationEndpointService);
        choices->Append(ref new DeviceInformationKindChoice("AssociationEndpointService", kinds->GetView()));

        kinds = ref new Vector<DeviceInformationKind>();
        kinds->Append(DeviceInformationKind::AssociationEndpointService);
        kinds->Append(DeviceInformationKind::DeviceInterface);
        choices->Append(ref new DeviceInformationKindChoice("AssociationEndpointService and DeviceInterface", kinds->GetView()));
        
        return choices->GetView();
    }

    IVectorView<ProtectionLevelSelectorInfo^>^ ProtectionSelectorChoices::Selectors::get()
    {
        Vector<ProtectionLevelSelectorInfo^>^ selectors = ref new Vector<ProtectionLevelSelectorInfo^>();

        selectors->Append(ref new ProtectionLevelSelectorInfo("Default", Windows::Devices::Enumeration::DevicePairingProtectionLevel::Default ));
        selectors->Append(ref new ProtectionLevelSelectorInfo("None", Windows::Devices::Enumeration::DevicePairingProtectionLevel::None ));
        selectors->Append(ref new ProtectionLevelSelectorInfo("Encryption", Windows::Devices::Enumeration::DevicePairingProtectionLevel::Encryption ));
        selectors->Append(ref new ProtectionLevelSelectorInfo("Encryption and authentication", Windows::Devices::Enumeration::DevicePairingProtectionLevel::EncryptionAndAuthentication ));

        return selectors->GetView();
    }

    // DeviceInformationDisplay

    DeviceInformationDisplay::DeviceInformationDisplay(DeviceInformation^ deviceInfoIn)
    {
        deviceInfo = deviceInfoIn;
        UpdateGlyphBitmapImage();
    }

    String^ DeviceInformationDisplay::KindString::get()
    {
        String^ str;
        switch (deviceInfo->Kind)
        {
        case DeviceInformationKind::DeviceContainer:
            str = "DeviceContainer";
            break;
        case DeviceInformationKind::Device:
            str = "Device";
            break;
        case DeviceInformationKind::DeviceInterface:
            str = "DeviceInterface";
            break;
        case DeviceInformationKind::DeviceInterfaceClass:
            str = "DeviceInterfaceClass";
            break;
        case DeviceInformationKind::AssociationEndpointContainer:
            str = "AssociationEndpointContainer";
            break;
        case DeviceInformationKind::AssociationEndpoint:
            str = "AssociationEndpoint";
            break;
        case DeviceInformationKind::AssociationEndpointService:
            str = "AssociationEndpointService";
            break;
        };

        return str;
    }

    void DeviceInformationDisplay::Update(DeviceInformationUpdate^ deviceInfoUpdateIn)
    {
        deviceInfo->Update(deviceInfoUpdateIn);

        OnPropertyChanged("Kind");
        OnPropertyChanged("Id");
        OnPropertyChanged("Name");
        OnPropertyChanged("DeviceInformation");
        OnPropertyChanged("CanPair");
        OnPropertyChanged("IsPaired");

        UpdateGlyphBitmapImage();
    }

    void DeviceInformationDisplay::UpdateGlyphBitmapImage()
    {
        BitmapImage^ glyphBitmapImage;
        glyphBitmapImage = ref new BitmapImage();

        concurrency::create_task(deviceInfo->GetGlyphThumbnailAsync()).then(
            [glyphBitmapImage](DeviceThumbnail^ result)
        {
            return glyphBitmapImage->SetSourceAsync(result);
        }).then(
            [this, glyphBitmapImage]()
        {
            GlyphBitmapImage = glyphBitmapImage;
            OnPropertyChanged("GlyphBitmapImage");
        });
    }

    void DeviceInformationDisplay::OnPropertyChanged(String^ name)
    {
        MainPage::Current->Dispatcher->RunAsync(CoreDispatcherPriority::Normal, ref new DispatchedHandler(
            [this, name]()
        {
            PropertyChanged(this, ref new PropertyChangedEventArgs(name));
        }));
    }


    // GeneralPropertyValueConverter
    
    Object^ GeneralPropertyValueConverter::Convert(
        Object^ value,
        TypeName targetType,
        Object^ parameter,
        String^ language)
    {
        Object^ property = nullptr;
        String^ strParameter = safe_cast<String^>(parameter);

        if (nullptr != strParameter && !strParameter->IsEmpty())
        {
            IMapView<String^, Object^>^ properties = safe_cast<IMapView<String^, Object^>^>(value);

            property = properties->Lookup(strParameter);
        }

        return property;
    }
}

File: Samples/DeviceEnumerationAndPairing/cs/DisplayHelpers.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Generic;
using System.ComponentModel;
using Windows.Devices.Enumeration;
using Windows.UI.Xaml.Data;
using Windows.UI.Xaml.Media.Imaging;

// NOTE: The following using statements are only needed in order to demonstrate many of the
// different device selectors available from Windows Runtime APIs. You will only need to include
// the namespace for the Windows Runtime API your actual scenario needs.
using Windows.Devices.HumanInterfaceDevice;
using Windows.Networking.Proximity;
using Windows.Devices.Bluetooth;
using Windows.Devices.WiFiDirect;
using Windows.Media.Casting;
using Windows.Media.DialProtocol;
using Windows.Devices.Sensors;
using Windows.Devices.PointOfService;

namespace SDKTemplate
{
    public class DeviceSelectorInfo
    {
        public string DisplayName { get; set; }
        public DeviceClass DeviceClassSelector { get; set; } = DeviceClass.All;
        public DeviceInformationKind Kind { get; set; } = DeviceInformationKind.Unknown;
        public string Selector { get; set; }
    }

    public static class DeviceSelectorChoices
    {

        public static List<DeviceSelectorInfo> CommonDeviceSelectors => new List<DeviceSelectorInfo>()
        {
            // Pre-canned device class selectors
            new DeviceSelectorInfo() { DisplayName = "All Device Interfaces (default)", DeviceClassSelector = DeviceClass.All, Selector = null },
            new DeviceSelectorInfo() { DisplayName = "Audio Capture", DeviceClassSelector = DeviceClass.AudioCapture, Selector = null },
            new DeviceSelectorInfo() { DisplayName = "Audio Render", DeviceClassSelector = DeviceClass.AudioRender, Selector = null },
            new DeviceSelectorInfo() { DisplayName = "Image Scanner", DeviceClassSelector = DeviceClass.ImageScanner, Selector = null },
            new DeviceSelectorInfo() { DisplayName = "Location", DeviceClassSelector = DeviceClass.Location, Selector = null },
            new DeviceSelectorInfo() { DisplayName = "Portable Storage", DeviceClassSelector = DeviceClass.PortableStorageDevice, Selector = null },
            new DeviceSelectorInfo() { DisplayName = "Video Capture", DeviceClassSelector = DeviceClass.VideoCapture, Selector = null },

            // A few examples of selectors built dynamically by windows runtime apis.
            new DeviceSelectorInfo() { DisplayName = "Human Interface (HID)", Selector = HidDevice.GetDeviceSelector(0, 0) },
            new DeviceSelectorInfo() { DisplayName = "Activity Sensor", Selector = ActivitySensor.GetDeviceSelector() },
            new DeviceSelectorInfo() { DisplayName = "Pedometer", Selector = Pedometer.GetDeviceSelector() },
            new DeviceSelectorInfo() { DisplayName = "Proximity", Selector = ProximityDevice.GetDeviceSelector() },
            new DeviceSelectorInfo() { DisplayName = "Proximity Sensor", Selector = ProximitySensor.GetDeviceSelector() },
        };

        // Currently Bluetooth APIs don't provide a selector to get ALL devices that are both paired and non-paired.
        // Typically you wouldn't need this for common scenarios, but it's convenient to demonstrate the
        // various sample scenarios.
        public static DeviceSelectorInfo Bluetooth =>
            new DeviceSelectorInfo() { DisplayName = "Bluetooth", Selector = "System.Devices.Aep.ProtocolId:=\"{e0cbf06c-cd8b-4647-bb8a-263b43f0f974}\"", Kind = DeviceInformationKind.AssociationEndpoint };

        public static DeviceSelectorInfo BluetoothUnpairedOnly =>
            new DeviceSelectorInfo() { DisplayName = "Bluetooth (unpaired)", Selector = BluetoothDevice.GetDeviceSelectorFromPairingState(false) };

        public static DeviceSelectorInfo BluetoothPairedOnly =>
            new DeviceSelectorInfo() { DisplayName = "Bluetooth (paired)", Selector = BluetoothDevice.GetDeviceSelectorFromPairingState(true) };

        // Currently Bluetooth APIs don't provide a selector to get ALL devices that are both paired and non-paired.
        // Typically you wouldn't need this for common scenarios, but it's convenient to demonstrate the
        // various sample scenarios.
        public static DeviceSelectorInfo BluetoothLE =>
            new DeviceSelectorInfo() { DisplayName = "Bluetooth LE", Selector = "System.Devices.Aep.ProtocolId:=\"{bb7bb05e-5972-42b5-94fc-76eaa7084d49}\"", Kind = DeviceInformationKind.AssociationEndpoint };

        public static DeviceSelectorInfo BluetoothLEUnpairedOnly =>
            new DeviceSelectorInfo() { DisplayName = "Bluetooth LE (unpaired)", Selector = BluetoothLEDevice.GetDeviceSelectorFromPairingState(false) };

        public static DeviceSelectorInfo BluetoothLEPairedOnly =>
            new DeviceSelectorInfo() { DisplayName = "Bluetooth LE (paired)", Selector = BluetoothLEDevice.GetDeviceSelectorFromPairingState(true) };

        public static DeviceSelectorInfo WiFiDirect =>
            new DeviceSelectorInfo() { DisplayName = "Wi-Fi Direct", Selector = WiFiDirectDevice.GetDeviceSelector(WiFiDirectDeviceSelectorType.AssociationEndpoint) };

        public static DeviceSelectorInfo WiFiDirectPairedOnly =>
            new DeviceSelectorInfo() { DisplayName = "Wi-Fi Direct (paired)", Selector = WiFiDirectDevice.GetDeviceSelector() };

        public static DeviceSelectorInfo PointOfServicePrinter =>
            new DeviceSelectorInfo() { DisplayName = "Point of Service Printer", Selector = PosPrinter.GetDeviceSelector() };

        public static DeviceSelectorInfo VideoCasting =>
            new DeviceSelectorInfo() { DisplayName = "Video Casting", Selector = CastingDevice.GetDeviceSelector(CastingPlaybackTypes.Video) };

        public static DeviceSelectorInfo DialAllApps =>
            new DeviceSelectorInfo() { DisplayName = "DIAL (All apps)", Selector = DialDevice.GetDeviceSelector("") };

        // WSD and UPnP are unique in that there are currently no general WSD or UPnP APIs to USE the devices once you've discovered them.
        // You can pair the devices using DeviceInformation.Pairing.PairAsync etc, and you can USE them with the sockets APIs. However, since
        // there's no specific API right now, there's no *.GetDeviceSelector available.  That's why we just simply build the device selector
        // ourselves and specify the correct DeviceInformationKind (AssociationEndpoint).
        public static DeviceSelectorInfo Wsd =>
            new DeviceSelectorInfo() { DisplayName = "Web Services on Devices (WSD)", Selector = "System.Devices.Aep.ProtocolId:=\"{782232aa-a2f9-4993-971b-aedc551346b0}\"", Kind = DeviceInformationKind.AssociationEndpoint };

        public static DeviceSelectorInfo Upnp =>
            new DeviceSelectorInfo() { DisplayName = "UPnP", Selector = "System.Devices.Aep.ProtocolId:=\"{0e261de4-12f0-46e6-91ba-428607ccef64}\"", Kind = DeviceInformationKind.AssociationEndpoint };

        public static DeviceSelectorInfo NetworkCamera =>
            new DeviceSelectorInfo() { DisplayName = "Web Services on Devices (NetworkCamera)", Selector = "System.Devices.Aep.ProtocolId:=\"{43cc0de4-8ca8-4a56-805a-86fc63f21602}\"", Kind = DeviceInformationKind.AssociationEndpoint };

        public static void AddVideoCastingIfSupported(List<DeviceSelectorInfo> selectors)
        {
            try
            {
                selectors.Add(VideoCasting);
            }
            catch (Exception ex) when (ex.HResult == unchecked((int)0x80070032)) // HRESULT_FROM_WIN32(ERROR_NOT_SUPPORTED)
            {
                // Video casting is not supported by the system.
            }
        }

        public static List<DeviceSelectorInfo> DevicePickerSelectors
        {
            get
            {
                // Add all selectors that can be used with the DevicePicker
                List<DeviceSelectorInfo> selectors = new List<DeviceSelectorInfo>(CommonDeviceSelectors);
                selectors.Add(BluetoothPairedOnly);
                selectors.Add(BluetoothUnpairedOnly);
                selectors.Add(BluetoothLEPairedOnly);
                selectors.Add(BluetoothLEUnpairedOnly);
                selectors.Add(WiFiDirect);
                selectors.Add(PointOfServicePrinter);
                AddVideoCastingIfSupported(selectors);
                selectors.Add(DialAllApps);

                return selectors;
            }
        }

        // Add all selectors that are reasonable to use with FindAllAsync
        public static List<DeviceSelectorInfo> FindAllAsyncSelectors =>
            new List<DeviceSelectorInfo>(CommonDeviceSelectors)
            {
                BluetoothPairedOnly,
                BluetoothLEPairedOnly,
                WiFiDirectPairedOnly
            };

        public static List<DeviceSelectorInfo> DeviceWatcherSelectors
        {
            get
            {
                // Add all selectors that can be used with the DeviceWatcher
                List<DeviceSelectorInfo> selectors = new List<DeviceSelectorInfo>(CommonDeviceSelectors);
                selectors.Add(Bluetooth);
                selectors.Add(BluetoothLE);
                selectors.Add(WiFiDirect);
                selectors.Add(PointOfServicePrinter);
                AddVideoCastingIfSupported(selectors);
                selectors.Add(DialAllApps);
                selectors.Add(Wsd);
                selectors.Add(Upnp);

                return selectors;
            }
        }

        public static List<DeviceSelectorInfo> BackgroundDeviceWatcherSelectors
        {
            get
            {
                // Add all selectors that can be used with the BackgroundDeviceWatcher
                List<DeviceSelectorInfo> selectors = new List<DeviceSelectorInfo>(CommonDeviceSelectors);
                selectors.Add(BluetoothPairedOnly);
                selectors.Add(BluetoothLEPairedOnly);
                selectors.Add(WiFiDirectPairedOnly);
                selectors.Add(PointOfServicePrinter);
                AddVideoCastingIfSupported(selectors);
                selectors.Add(DialAllApps);
                selectors.Add(Wsd);
                selectors.Add(Upnp);

                return selectors;
            }
        }

        /// <summary>
        /// Selectors for use in the pairing scenarios
        /// </summary>
        public static List<DeviceSelectorInfo> PairingSelectors
        {
            get
            {
                // Add selectors that can be used in pairing scenarios
                List<DeviceSelectorInfo> selectors = new List<DeviceSelectorInfo>();
                selectors.Add(Bluetooth);
                selectors.Add(BluetoothLE);
                selectors.Add(WiFiDirect);
                selectors.Add(PointOfServicePrinter);
                AddVideoCastingIfSupported(selectors);
                selectors.Add(Wsd);
                selectors.Add(Upnp);
                selectors.Add(NetworkCamera);

                return selectors;
            }
        }
    }

    public struct DeviceInformationKindChoice
    {
        public string DisplayName { get; set; }
        public DeviceInformationKind[] DeviceInformationKinds { get; set; }
    }

    public static class DeviceInformationKindChoices
    {
        public static List<DeviceInformationKindChoice> Choices =>
            new List<DeviceInformationKindChoice>()
            {
                new DeviceInformationKindChoice() { DisplayName = "DeviceContainer", DeviceInformationKinds = new[] { DeviceInformationKind.DeviceContainer } },
                new DeviceInformationKindChoice() { DisplayName = "Device", DeviceInformationKinds = new[] { DeviceInformationKind.Device } },
                new DeviceInformationKindChoice() { DisplayName = "DeviceInterface (Default)", DeviceInformationKinds = new[] { DeviceInformationKind.DeviceInterface } },
                new DeviceInformationKindChoice() { DisplayName = "DeviceInterfaceClass", DeviceInformationKinds = new[] { DeviceInformationKind.DeviceInterfaceClass } },
                new DeviceInformationKindChoice() { DisplayName = "AssociationEndpointContainer", DeviceInformationKinds = new[] { DeviceInformationKind.AssociationEndpointContainer } },
                new DeviceInformationKindChoice() { DisplayName = "AssociationEndpoint", DeviceInformationKinds = new[] { DeviceInformationKind.AssociationEndpoint } },
                new DeviceInformationKindChoice() { DisplayName = "AssociationEndpointService", DeviceInformationKinds = new[] { DeviceInformationKind.AssociationEndpointService } },
                new DeviceInformationKindChoice() { DisplayName = "AssociationEndpointService and DeviceInterface", DeviceInformationKinds = new[] { DeviceInformationKind.AssociationEndpointService, DeviceInformationKind.DeviceInterface } },
            };
    }

    public struct ProtectionLevelSelectorInfo
    {
        public string DisplayName { get; set; }
        public DevicePairingProtectionLevel ProtectionLevel { get; set; }
    }

    public static class ProtectionSelectorChoices
    {
        public static List<ProtectionLevelSelectorInfo> Selectors =>
            new List<ProtectionLevelSelectorInfo>()
            {
                new ProtectionLevelSelectorInfo() { DisplayName = "Default", ProtectionLevel = DevicePairingProtectionLevel.Default },
                new ProtectionLevelSelectorInfo() { DisplayName = "None", ProtectionLevel = DevicePairingProtectionLevel.None },
                new ProtectionLevelSelectorInfo() { DisplayName = "Encryption", ProtectionLevel = DevicePairingProtectionLevel.Encryption },
                new ProtectionLevelSelectorInfo() { DisplayName = "Encryption and authentication", ProtectionLevel = DevicePairingProtectionLevel.EncryptionAndAuthentication },
            };
    }

    public class DeviceInformationDisplay : INotifyPropertyChanged
    {
        public DeviceInformationDisplay(DeviceInformation deviceInfoIn)
        {
            DeviceInformation = deviceInfoIn;
            UpdateGlyphBitmapImage();
        }

        public DeviceInformationKind Kind => DeviceInformation.Kind;
        public string Id => DeviceInformation.Id;
        public string Name => DeviceInformation.Name;
        public BitmapImage GlyphBitmapImage { get; private set; }
        public bool CanPair => DeviceInformation.Pairing.CanPair;
        public bool IsPaired => DeviceInformation.Pairing.IsPaired;
        public IReadOnlyDictionary<string, object> Properties => DeviceInformation.Properties;
        public DeviceInformation DeviceInformation { get; private set; }

        public void Update(DeviceInformationUpdate deviceInfoUpdate)
        {
            DeviceInformation.Update(deviceInfoUpdate);

            OnPropertyChanged("Kind");
            OnPropertyChanged("Id");
            OnPropertyChanged("Name");
            OnPropertyChanged("DeviceInformation");
            OnPropertyChanged("CanPair");
            OnPropertyChanged("IsPaired");
            OnPropertyChanged("GetPropertyForDisplay");

            UpdateGlyphBitmapImage();
        }

        public string GetPropertyForDisplay(string key) => Properties[key]?.ToString();

        private async void UpdateGlyphBitmapImage()
        {
            DeviceThumbnail deviceThumbnail = await DeviceInformation.GetGlyphThumbnailAsync();
            BitmapImage glyphBitmapImage = new BitmapImage();
            await glyphBitmapImage.SetSourceAsync(deviceThumbnail);
            GlyphBitmapImage = glyphBitmapImage;
            OnPropertyChanged("GlyphBitmapImage");
        }

        public event PropertyChangedEventHandler PropertyChanged;
        protected void OnPropertyChanged(string name) => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
    }
}
